### PR TITLE
Pattern match

### PR DESCRIPTION
**Pull request by MajronMan**

### CODE REVIEW

File: src/Ast/Pattern.elm [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            choice
+                [ wildParser
+                , varParser
+                , constructorParser
+                , PLiteral <$> literalParser
+                , PRecord <$> (braces <| commaSeparated_ loName)
+                , PTuple <$> tupleParser pattern
+                , PList <$> listParser pattern
+                , parens pattern
+                ]
```
Comment (wende) [inline src/Ast/Pattern.elm:114]: This definitely needs a refactor. I'll give a moment a bit later

File: tests/Helpers.elm [hunk 2]
```diff
@@ -1,9 +1,11 @@
-module Helpers exposing (ExpressionSansMeta(..), StatementSansMeta(..), access, accessFun, app, areStatements, areStatementsSansMeta, binOp, case_, character, characterPattern, comment, dropDoubleMExp, dropExpressionMeta, dropMExpMeta, dropStatementMeta, dropWithMetaMExp, effectModuleDeclaration, fails, failsPattern, failure, float, floatPattern, functionDeclaration, functionTypeDeclaration, importStatement, infixDeclaration, integer, integerPattern, isApplicationSansMeta, isExpression, isExpressionSansMeta, isPattern, isStatement, isStatementSansMeta, lambda, let_, list, moduleDeclaration, portDeclaration, portModuleDeclaration, portTypeDeclaration, record, recordUpdate, simpleParse, string, stringPattern, tuple, typeAliasDeclaration, typeDeclaration, var)
+module Helpers exposing (..)
```
Comment (wende) [inline tests/Helpers.elm]: Hehe

File: tests/Pattern.elm [hunk 3]
```diff
@@ -73,15 +73,42 @@ constructor =
...
+                                            (floatPattern 2.5)
+                                        )
+                                        (characterPattern 'a')
+                                    )
+                                    (stringPattern "b")
+                                )
+                                PWild
+                            )
+                            (PVariable "c")
+                        )
```
Comment (wende) [inline tests/Pattern.elm:90]: Can be done much nicer with fold like here:
https://github.com/Bogdanp/elm-ast/blob/0e8caebf225328ee0c93d3d8bb0ba354f4dfbc43/tests/Helpers.elm#L369

File: tests/Pattern.elm [hunk 4]
```diff
@@ -73,15 +73,42 @@ constructor =
...
                 "A (B 2) (C 3 4)"
-                    |> isPattern (PConstructor "A" [ PConstructor "B" [ integerPattern 2 ], PConstructor "C" [ integerPattern 3, integerPattern 4 ] ])
+                    |> isPattern
+                        (PApplication
+                            (PApplication
+                                (PConstructor "A")
+                                (PApplication (PConstructor "B") (integerPattern 2))
+                            )
+                            (PApplication (PApplication (PConstructor "C") (integerPattern 3)) (integerPattern 4))
+                        )
```
Comment (wende) [inline tests/Pattern.elm:101]: ditto

File: tests/Pattern.elm [hunk 5]
```diff
@@ -129,18 +156,70 @@ cons =
...
+        [ test "simple" <|
+            \() ->
+                "f x y _"
+                    |> isPattern
+                        (PApplication
+                            (PApplication
+                                (PApplication (PVariable "f") (PVariable "x"))
+                                (PVariable "y")
+                            )
+                            PWild
```
Comment (wende) [inline tests/Pattern.elm]: ditto

File: tests/Pattern.elm [hunk 6]
```diff
@@ -129,18 +156,70 @@ cons =
...
+                        (PApplication
+                            (PApplication
+                                (PApplication
+                                    (PApplication (PVariable "a") (PConstructor "B"))
+                                    (PVariable "c")
+                                )
+                                (PConstructor "D")
+                            )
+                            (PVariable "e")
+                        )
```
Comment (wende) [inline tests/Pattern.elm:197]: ditto

File: tests/Pattern.elm [hunk 7]
```diff
@@ -129,18 +156,70 @@ cons =
...
+                                    (PApplication
+                                        (PApplication (PVariable "f") (PApplication (PConstructor "Just") (PVariable "x")))
+                                        (PConstructor "Nothing")
+                                    )
+                                    (PVariable "y")
+                                )
+                                (PAs (PCons (PVariable "a") (PVariable "b")) "c")
+                            )
+                            (PTuple [ PWild, PLiteral <| Float 3.6 ])
+                        )
```
Comment (wende) [inline tests/Pattern.elm:208]: ditto

File: tests/Statement.elm [hunk 8]
```diff
@@ -354,8 +360,10 @@ multipleDeclarations =
                             (TypeTuple [ TypeConstructor [ "Int" ] [], TypeConstructor [ "Int" ] [] ])
                             (TypeConstructor [ "Int" ] [])
                         )
-                    , functionDeclaration "h"
-                        [ tuple [ var "a", var "b" ] ]
+                    , functionDeclaration
+                        (applicationPattern (variablePattern "h")
+                            (tuplePattern [ variablePattern "a", variablePattern "b" ])
+                        )
```
Comment (wende) [inline tests/Statement.elm:366]: ditto (?)

File: tests/Statement.elm [hunk 9]
```diff
@@ -366,7 +374,12 @@ multipleDeclarations =
                             (TypeConstructor [ "Int" ] [])
                             (TypeConstructor [ "Int" ] [])
                         )
-                    , functionDeclaration "+" [ var "a", var "b" ] (integer 1)
+                    , functionDeclaration
+                        (applicationPattern
+                            (applicationPattern (variablePattern "+") (variablePattern "a"))
+                            (variablePattern "b")
+                        )
+                        (integer 1)
```
Comment (wende) [inline tests/Statement.elm:382]: Same

**Review by wende — CHANGES_REQUESTED** [review]

File: tests/Statement.elm (hunk 8 above)
Comment (MajronMan) [reply tests/Statement.elm:366]: Won't help with less than 3 args

File: tests/Statement.elm (hunk 9 above)
Comment (MajronMan) [reply tests/Statement.elm:382]: Also won't help

**Pull request merged**